keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,350 @@
+-->
+
+### Risks and Mitigations
+
+If the workload fails to schedule its pods it could block admission of other
+workloads indefinitely.
+
+In order to mitigate this risk we terminate workloads which don't have enough
+pods ready after the timeout.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### Ready workload condition
+
+First, we introduce a new workload condition, called `Ready`, to denote workloads
+which have their pods scheduled, more precisely, which satisfy the condition:
+`job.status.ready + job.status.succeeded >= job.spec.parallelism`.
+The condition is added by the kueue's Job controller when it reconciles a job on
+a Job status update. Note that, verifying the condition does not require extra
+API calls to the job or individual pods.
+
+### Workload API
+
+We extend the Workload API in order to specify the strategy
+
+```golang
+// WorkloadSpec defines the desired state of Workload
+type WorkloadSpec struct {
+  ...
+  // waitForPodsScheduled is true indicates that the workload blocks admission
+  // of all other workloads until it is in the Ready condition. Further, it
+  // delays its corresponding job start until all other admitted workloads
+  // are in the Ready condition. If false, then the workloads starts once
+  // admitted. Also, it does not block admission of other workloads. The null
+  // value, results in the same behavior as false.
+	waitForPodsScheduled *bool `json:"schedulingStrategy,omitempty"`
+}
+```
+
+### Job annotations
+
+We use the mechanism of job annotations to translate them into the desired
+workload properties.
+
+#### Wait for pod scheduling
+
+We introduce a new Job annotation, called `kueue.x-k8s.io/wait-for-pod-scheduling`,
+to indicate that the workload corresponing to this job should block admission
+of other workloads until it is scheduled. Further, workloads with this annotation
+wait with unsuspending the corresponding job until all other admitted workloads
+have their pods scheduled.
+
+#### Pod scheduled timeout seconds
+
+We introduce a new annotation, called `kueue.x-k8s.io/pods-scheduled-timeout-seconds`,
+to terminate any workloads which don't have their pods scheduled within the
+specified time since unsuspending of the corresponding job.
+
+If the annotation is not specified the timeout fallbacks to `1s` per pod + `30s`
```

Comment:
Maybe by default there is no timeout? Not sure.

Reply:
Yes, I'm hesitant too. It is also sort of a problem today, so maybe it can be done but decoupled from this KEP?